doc/specs/stdlib_hash.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Hashing algorithms
```

Comment:
@wclodius2 was working on an implementation of several hash procedures. It is maybe worthwhile to check with him what is the status of this project.

Reply:
I was actually planning to just use the C implementation until we are able to replace it with @wclodius2 implementation. But than I had some issues with CMake and the manual makefile and decided it would be easier to port the C implementation directly to Fortran rather than waste my time on build systems.

If you think we should split this PR to discuss the hash implementation and the map base class separately, I close this one and create two separate PRs.

Also, the waterhash implementation is now available as standalone fpm project to ease the code review: https://github.com/awvwgk/waterhash.

Reply:
I ran into the problem that there are at least three general API's for hash functions:
1. A single object to be hashed as the single argument.
2. The hash object and a single salt (seed) value.
3. The hash object and two salt values. 
These different APIs reflect different goals for the hash functions. Those with the single argument are typically simple functions intended to hash small objects in situations where denial of service attacks are a negligible concern. I suspect they are appropriate for the majority of hash table applications. Those with a salt/seed are typically complicated functions optimized for hashing large objects, and use the seeds to deal with denial of service attacks. This is useful for internet applications, but I suspect less useful for the applications Fortran would be used for. Then there is the difference between 32 bit and 64 biT hashes. The Waterhash I know is aa 32 bit hash with aa single seed.

The design of the hash table will reflect the the type of hash function it is intended to use. If the hash function uses seeds the hash table will need to set aside storage for the seed, and have an argument representing the seed. Between the three APIs and 32 vs. 64 bits to address all combinations requires 6 different hash table implementations. I can't see doing that as it would confuse the users. I can see implementing two hash tables, so a decision should be made as to .

Reply:
I think we can and should provide several implementation for the map base class. We will probably need at least two specific implementations to cover unordered and ordered maps. In some applications I also have the need to preserve the order in which the elements are added to the map, even if it adds a runtime penality.

For very specific requirement of the map an external implementation of stdlib's base class could be used seamlessly as well if we design the base class well.

Reply:
I would propose that we start to implement something that would cover e.g., 90-95% of the cases we would meet in Fortran. IMO, it would be 64bit + single seed (32bit might be enough for most users but I usually need a 64bit  ;) ).
@wclodius2 I believe you already implemented and tested different hash functions. Could it be a start for such a module? @awvwgk could possibly complement it with is waterhash implementation. I also have a Fortran version of Lookup3. What do you think both?

Reply:
Agree, I see some value to start with a `stdlib_hash` module on its own featuring different hashing algorithms. I can start a separate PR to create such a module and we can start collecting hashing algorithms in Fortran there, which are used to implement the unordered maps later.

Reply:
I opened #455 to start the `stdlib_hash` module.

Reply:
I am going on a trip tomorrow. Preparing for the trip has kept me busy, and the first few days of the trip will be very busy. It will be difficult for me to do much until this weekend and maybe not until sometime next week. FWIW I have about a dozen hash code implementations, including one for Waterhash. I have a couple of hash table implementations that could be the basis of an unordered map. I had been fiddling with the hash tables as I tried out new hash codes so they may need some polishing before I share them. I also have a red-black tree implementation that could be the basis of an ordered map. Typically in my implementations the key and data are passed as `INT8` vectors, that are presumably generated by using transfer on the original data. The ordering in the red-black tree is therefore not what might be expected from the original data. This raises the following questions:
1. Is there interest in seeing the codes, and which codes have the highest interest?
2. What is the best way of distributing the code? As a proposed addition to `stdlib`, as a separate git project, or as an fpm project?
3. If it is to be distributed as a fpm project how do I go about doing that?

Reply:
Thank you @wclodius2 for your answer.

> I am going on a trip tomorrow. Preparing for the trip has kept me busy, and the first few days of the trip will be very busy. It will be difficult for me to do much until this weekend and maybe not until sometime next week. 

No problem. Enjoy your trip ;)

> 1. Is there interest in seeing the codes, and which codes have the highest interest?
Yes, sure. See #455 for example. 

> 2. What is the best way of distributing the code? As a proposed addition to `stdlib`, as a separate git project, or as an fpm project?

Since @awvwgk started #455, I guess that the best way is to combine our effort in #455, to discuss there API and implementations.

Reply:
So should I branch #455 and add my 32 bit hash codes as a PR to it, or should I start my own FPM project, or something else?

Reply:
I suggest that you start from #455. @awvwgk is ok for you?

Reply:
Whatever works best for you.